src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -516,21 +520,66 @@ private static Expression GetValueFromProperty(Expression sourceExpression, stri
             return Expression.Convert(indexExpression, typeof(string));
         }
 
+        private static Expression BindParameterFromService(ParameterInfo parameter)
+        {
+            var nullability = NullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState == NullabilityState.Nullable;
+
+            return isOptional
+                ? Expression.Call(GetServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr)
+                : Expression.Call(GetRequiredServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr);
+        }
+
         private static Expression BindParameterFromValue(ParameterInfo parameter, Expression valueExpression, FactoryContext factoryContext)
         {
+            var nullability = NullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState == NullabilityState.Nullable;
+
+            var argument = Expression.Variable(parameter.ParameterType, $"{parameter.Name}_local");
+
             if (parameter.ParameterType == typeof(string))
             {
-                if (!parameter.HasDefaultValue)
+                if (!isOptional)
+                {
+                    // The following is produced if the parameter is required:
+                    //
+                    // tempSourceString = httpContext.RouteValue["param1"] ?? httpContext.Query["param1"];
+                    // if (tempSourceString == null)
+                    // {
+                    //      wasTryParseFailure = true;
```

Comment:
Let's rename this to something more generic like `wasParamCheckFailure` or something. And we should rename `WasTryParseFailureExpr` and anything like that as well.